### PR TITLE
More streamlining of model interface and train scripts

### PR DESCRIPTION
**Pull request by vinamarora8**
The PR brings in _three_ features/ideas.  Currently these are only applied to POYO and its surrounding code. If/once we are happy with these, we can update POYO+ with similar ideas.

1. Tokenizer is made part of the Model.
    - I am fairly confident that putting tokenizer in model does not copy the model's weight tensors (which are on GPU) to the workers. I printed `id(self.unit_emb.weight)` inside the tokenizer, and all workers had the same id. This should mean `unit_emb.weight` points to the same place in memory for all workers and is not a fresh copy for each.
    - `DataModule.setup()` has been changed to `DataModule.setup_dataset_and_link_model(model: POYO)` which sets up the train/val/test datasets, links the tokenize fn to the datasets, and updates the embedding vocabularies. I think this is a good way to express the tight relationship between our data and model, but open to suggestions/criticisms.
    
2. Tokenizer's output now has a `batch["model_input"]` which can straightaway be pass to the model. No popping items from batch dictionary required.
    - I am not super happy about accessing output mask through `batch["model_input"]["output_mask"]`, but the only alternative I see is to have another copy of this as `batch["output_mask"]` which has (very slight, but non-zero in general) dataloader performance implications.
    
3. Data interface to `DecodingStitchEvaluator` is formalized with a dataclass: `DataForDecodingStitchEvaluator`
    - This way the Callback based StitchEvaluator framework becomes a bit more transparent, and more generally usable by non-POYO train scripts. The problem with current approach was the requirement to follow a batch dictionary format that was hidden in documentation.
    - I know we might be considering changing StitchEvaluator API altogether. But this seems like a better version to keep in the main branch until we get to reworking the API.

#### TODOs for once we're happy about these ideas:
- [x] Mirror changes in POYO+ and MultiTaskDecodingStitchEvaluator
- [x] Tests
- [x] Doc Updates:
  - [x] POYO
  - [x] POYO+
  - [x] DecodingStitchEvaluator
  - [x] ~MultitaskDecodingStitchEvaluator~ (it never had a docstring, so this can be a separate PR)

### CODE REVIEW

File: torch_brain/models/poyo.py [hunk 1]
```diff
@@ -337,7 +285,7 @@ def __call__(self, data: Data) -> Dict:
 
         # unit_index is relative to the recording, so we want it to map it to
         # the global unit index
-        local_to_global_map = np.array(self.unit_tokenizer(unit_ids))
+        local_to_global_map = np.array(self.unit_emb.tokenizer(unit_ids))
```
Comment (mazabou) [inline torch_brain/models/poyo.py:288]: note the inconsistency between `tokenizer` and `tokenize`

**Review by mazabou — COMMENTED** [review]
This is a move in the right direction! Overall the changes make sense to me. I just have a few nits here and there (sorry in advance).

File: examples/poyo/train.py [hunk 2]
```diff
@@ -283,25 +290,10 @@ def main(cfg: DictConfig):
...
-    data_module = DataModule(cfg=cfg, tokenizer=tokenizer)
-    data_module.setup()
-
-    # register units and sessions
-    model.unit_emb.initialize_vocab(data_module.get_unit_ids())
-    model.session_emb.initialize_vocab(data_module.get_session_ids())
+    # make model and dota module
+    model = poyo_mp(readout_spec=readout_spec)
+    data_module = DataModule(cfg=cfg)
+    data_module.setup_dataset_and_link_model(model)
```
[the remark this answers is not in the document]
Comment (mazabou) [reply examples/poyo/train.py:296]: you seem to have replaced the `setup`, not sure if it messes with things.

File: examples/poyo/train.py [hunk 3]
```diff
@@ -77,17 +81,15 @@ def configure_optimizers(self):
         }
 
     def training_step(self, batch, batch_idx):
-        target_values = batch.pop("target_values")
-        target_weights = batch.pop("target_weights")
-        output_mask = batch.pop("output_mask")
 
         # forward pass
-        output_values = self.model(**batch)
+        output_values = self.model(**batch["model_input"])
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:86]: idm "model_inputs"

File: examples/poyo/train.py [hunk 4]
```diff
@@ -109,73 +111,78 @@ def training_step(self, batch, batch_idx):
...
+            transform=partial(model.tokenize, eval_mode=True),
         )
         self.val_dataset.disable_data_leakage_check()
 
         self.test_dataset = Dataset(
             root=self.cfg.data_root,
             config=self.cfg.dataset,
             split="test",
-            transform=eval_tokenizer,
+            transform=partial(model.tokenize, eval_mode=True),
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:177]: I was thinking of doing it that way, but there's (very very small but non-zero) dataloader performance cost to sending those unneeded tensors to the GPU during training.

[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:177]: I'm open to eating that cost for the sake of cleanliness here.

File: examples/poyo/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:296]: Re: link model. It's a 2 way communication bw datasets and model. Model gets vocab update, and Dataset get the tokenizer from the model. So "link" seems natural. Other options could be 
- `setup_datasets_and_create_model_vocabs`
- `initialize_datasets_and_model_vocab`

File: torch_brain/models/poyo.py [hunk 5]
```diff
@@ -63,27 +65,38 @@ class POYO(nn.Module):
...
+        ffn_dropout: float = 0.2,
+        lin_dropout: float = 0.4,
+        atn_dropout: float = 0.0,
+        emb_init_scale: float = 0.02,
+        t_min: float = 1e-4,
+        t_max: float = 4.0,
     ):
         super().__init__()
 
+        self._ensure_valid_params(sequence_length, latent_step)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply torch_brain/models/poyo.py:86]: perhaps `_validate_params(...)`?

File: torch_brain/models/poyo.py (hunk 1 above)
Comment (vinamarora8) [reply torch_brain/models/poyo.py:288]: Yea I noticed that too. We're stuck between two non-ideal choices. 
1. `model.tokenize`: Function names are generally supposed to be "action phrases," which this is. But then we have to be okay with passing around functions as action-phrased named variables. 
2. `model.tokenizer`: It's not an action phrase, but makes more sense when being used as a variable.

If it means anything - ChatGPT seems to think `model.tokenize` is okay :3

File: examples/poyo/train.py [hunk 6]
```diff
@@ -109,73 +111,78 @@ def training_step(self, batch, batch_idx):
...
+        """
+        self.sequence_length = model.sequence_length
+
         train_transforms = hydra.utils.instantiate(self.cfg.train_transforms)
         self.train_dataset = Dataset(
             root=self.cfg.data_root,
             config=self.cfg.dataset,
             split="train",
-            transform=Compose([*train_transforms, self.tokenizer]),
+            transform=Compose([*train_transforms, model.tokenize]),
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:159]: Idk, we're stuck between two non-ideal choices. 

1. `model.tokenize`: Function names are generally supposed to be "action phrases," which this is. But then we have to be okay with passing around functions as action-phrase-named variables. 
2. `model.tokenizer`: It's not an action phrase, but makes more sense when being passed around as a variable.

I think a transform having an action-phrase name is okay. It's a function afterall.
If it means anything - ChatGPT too seems to think `model.tokenize` is okay :3

File: examples/poyo/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py]: done

File: examples/poyo/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:296]: Re: `setup`: It's a hook that's empty by default, users are supposed to define it.

File: examples/poyo/train.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mazabou) [reply examples/poyo/train.py:159]: ok we should make the change to InfiniteVocabEmbedding as well then.

File: examples/poyo/train.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mazabou) [reply examples/poyo/train.py:177]: yep, let's do it!

File: examples/poyo/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mazabou) [reply examples/poyo/train.py:296]: the hook is automatically called when you create a dataloader though, right?

File: torch_brain/models/poyo.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mazabou) [reply torch_brain/models/poyo.py:86]: yes, i like that more

File: torch_brain/models/poyo.py (hunk 1 above)
Comment (mazabou) [reply torch_brain/models/poyo.py:288]: ok, let's call it tokenize

Comment (vinamarora8) [reply torch_brain/models/poyo.py:288]: Changing `tokenizer` -> `tokenize` in `InfiniteVocabEmbedding` is going to touch a lot of files. I'll make it a separate PR instead.

File: examples/poyo/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo/train.py:296]: Resolved this conversation offline

**Comment by vinamarora8** [comment]
Thanks for the comments. Implementing the same for POYO+ before I ask for another review.

File: torch_brain/utils/readout.py [hunk 7]
```diff
@@ -77,7 +77,7 @@ def prepare_for_readout(
 
     # resolve eval mask
     eval_mask = None
-    eval_interval_key = data.config.get("eval_interval", None)
```
Comment (vinamarora8) [inline torch_brain/utils/readout.py:80]: This is off-topic from the PR description. But seems like a bug to me, so I fixed it here.

File: torch_brain/utils/readout.py [hunk 8]
```diff
@@ -77,7 +77,7 @@ def prepare_for_readout(
...
+    eval_interval_key = readout_config.get("eval_interval", None)
     if eval_interval_key is not None:
```
Comment (vinamarora8) [inline torch_brain/utils/readout.py:81]: This is off-topic for this PR. But seems like a bug to me, so I'm fixing it here.

**Comment by vinamarora8** [comment]
@mazabou I have mirrored the changes for POYO+, its train script, and its StitchEvaluator; and also updated `test_poyo_plus.py` to accommodate these edits. Ready for a final/second review.

File: torch_brain/utils/stitcher.py [hunk 9]
```diff
@@ -225,43 +239,58 @@ def on_test_epoch_end(self, *args, **kwargs):
...
+        self,
+        trainer: L.Trainer,
+        pl_module: L.LightningModule,
+        data: DataForMultiTaskDecodingStitchEvaluator,
+        *args,
+        **kwargs,
+    ):
         # update the cache with the predictions and targets
-        for readout_index in torch.unique(batch["output_decoder_index"]):
+        for readout_index in torch.unique(data.decoder_indices):
```
Comment (mazabou) [inline torch_brain/utils/stitcher.py:274]: I don't like that we call these `data`. `data` is used everywhere else to mean the `Data` object. Would you consider a different name?

Comment (mazabou) [reply torch_brain/utils/stitcher.py:274]: Actually, we will probably change this with the stitch PR, so we can ignore this.

**Review by mazabou — APPROVED** [review]
LGTM, thanks for this huge PR
